lib/urireader.cpp

```diff
@@ -97,12 +149,37 @@ namespace HTTP{
       }
     }
 
+    // In case of s3 URI we prepare HTTP request with AWS authorization and rely on HTTP logic below
+    if (myURI.protocol == "s3+https" || myURI.protocol == "s3+http"){
+      // Check fallback to global credentials in env vars
+      bool haveCredentials = myURI.user.size() && myURI.pass.size();
+      if(!haveCredentials) {
+        // Use environment variables
+        char * envValue = std::getenv("S3_ACCESS_KEY_ID");
+        if(envValue == NULL) {
+          FAIL_MSG("error s3 uri without credentials. Consider setting S3_ACCESS_KEY_ID env var");
+          return false;
+        }
+        myURI.user = envValue;
+        envValue = std::getenv("S3_SECRET_ACCESS_KEY");
+        if(envValue == NULL) {
+          FAIL_MSG("error s3 uri without credentials. Consider setting S3_SECRET_ACCESS_KEY env var");
+          return false;
+        }
+        myURI.pass = envValue;
+      }
+      // Do not return, continue to HTTP case
+      if(!transformS3ToHttp(downer, myURI)) return false;
+    } else {
+      // clear headers only if not s3 protocol
+      downer.clearHeaders();
```

Comment:
I think it might be better to move this clear up above the `if` block, since both the S3 and normal http(s) URIs use it. Easier to follow the logic, too, if there is only one `clearHeaders()` call and it happens unconditionally before the http-based handlers activate. 🤔